File: apps/doc-approval/windows-winui/DocApproval/ExecutionViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace DocApproval;

public partial class ExecutionViewModel : ObservableObject, IDisposable
{
    private readonly IEmbeddedHost? _host;
    private readonly ISettingsRepository _settings;
    private CancellationTokenSource? _submitCts;

    [ObservableProperty]
    private ExecutionPhase _phase = ExecutionPhase.Idle;

    [ObservableProperty]
    private string _document = string.Empty;

    [ObservableProperty]
    private RuntimeStatus _runtimeStatus = RuntimeStatus.Starting;

    [ObservableProperty]
    private bool _showTrace;

    [ObservableProperty]
    private string? _sessionId;

    [ObservableProperty]
    private DocApprovalOutput? _output;

    [ObservableProperty]
    private IReadOnlyList<TraceEvent> _trace = Array.Empty<TraceEvent>();

    [ObservableProperty]
    private string? _error;

    public ExecutionViewModel(IEmbeddedHost? host, ISettingsRepository settings)
    {
        _host = host;
        _settings = settings;
        RuntimeMode = EmbeddedHost.RuntimeModeEmbedded;
        WorkflowId = host?.WorkflowId ?? AppConstants.CapabilityId;
        RuntimeStatus = host?.IsReady == true ? RuntimeStatus.Ready : RuntimeStatus.Unavailable;
    }

    public string RuntimeMode { get; }

    public string WorkflowId { get; }

    public string Workspace => _settings.Workspace;

    public bool CanSubmit =>
        RuntimeStatus == RuntimeStatus.Ready &&
        !string.IsNullOrWhiteSpace(Document) &&
        Phase is not ExecutionPhase.Loading;

    partial void OnDocumentChanged(string value) => SubmitCommand.NotifyCanExecuteChanged();

    partial void OnPhaseChanged(ExecutionPhase value) => SubmitCommand.NotifyCanExecuteChanged();

    partial void OnRuntimeStatusChanged(RuntimeStatus value) => SubmitCommand.NotifyCanExecuteChanged();

    [RelayCommand(CanExecute = nameof(CanSubmit))]
    private async Task SubmitAsync()
    {
        if (!CanSubmit || _host is null)
        {
            return;
        }

        _submitCts?.Cancel();
        _submitCts = new CancellationTokenSource();

        Phase = ExecutionPhase.Loading;
        Error = null;
        Output = null;
        Trace = Array.Empty<TraceEvent>();
        ShowTrace = false;
        SessionId = null;

        var trimmedDocument = Document.Trim();

        try
        {
            var result = await Task.Run(
                () => _host.SubmitDocument(trimmedDocument),
                _submitCts.Token);

            SessionId = result.SessionId;
            Trace = result.Events;
            ShowTrace = result.Events.Count > 0;

            if (result.Error is not null)
            {
                Phase = ExecutionPhase.Failed;
                Error = result.Error;
                return;
            }

            Output = result.Output ?? DocApprovalOutput.Empty;
            Phase = ExecutionPhase.Succeeded;
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            Phase = ExecutionPhase.Failed;
            Error = ex.Message;
        }
    }

    [RelayCommand]
    private void Reset()
    {
        _submitCts?.Cancel();
        _submitCts = null;
        Phase = ExecutionPhase.Idle;
        SessionId = null;
        Output = null;
        Trace = Array.Empty<TraceEvent>();
        Error = null;
        ShowTrace = false;
    }

    public void RefreshRuntimeStatus()
    {
        RuntimeStatus = _host?.IsReady == true ? RuntimeStatus.Ready : RuntimeStatus.Unavailable;
        OnPropertyChanged(nameof(Workspace));
    }

    public void Dispose()
    {
        _submitCts?.Cancel();
        _submitCts?.Dispose();
        _host?.Dispose();
    }
}

File: apps/doc-approval/android-compose/app/src/main/java/com/traverseframework/docapproval/ExecutionViewModel.kt
package com.traverseframework.docapproval

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ExecutionViewModel(
    private val host: DocApprovalHost,
    private val settings: RuntimeSettings,
    private val computeDispatcher: CoroutineDispatcher = Dispatchers.Default,
) : ViewModel() {
    private val _uiState = MutableStateFlow(
        ExecutionUiState(
            runtimeStatus = if (host.isReady) RuntimeStatus.Ready else RuntimeStatus.Unavailable,
            runtimeMode = host.runtimeMode,
            workspace = AppConstants.DEFAULT_WORKSPACE,
        ),
    )
    val uiState: StateFlow<ExecutionUiState> = _uiState.asStateFlow()

    private var submitJob: Job? = null

    init {
        viewModelScope.launch {
            settings.workspace.collect { workspace ->
                _uiState.update { it.copy(workspace = workspace) }
            }
        }
    }

    fun updateDocument(document: String) {
        _uiState.update { it.copy(document = document.take(AppConstants.DOCUMENT_MAX_LENGTH)) }
    }

    fun toggleTrace(show: Boolean) {
        _uiState.update { it.copy(showTrace = show) }
    }

    fun submit() {
        val state = _uiState.value
        if (!state.canSubmit) return
        submitJob?.cancel()
        _uiState.update { it.copy(phase = ExecutionPhase.Loading) }
        val document = state.document.trim()
        submitJob = viewModelScope.launch {
            val result = withContext(computeDispatcher) { host.submitDocument(document) }
            if (result.error != null && result.output == null) {
                _uiState.update { it.copy(phase = ExecutionPhase.Failed(result.error)) }
            } else {
                _uiState.update {
                    it.copy(
                        phase = ExecutionPhase.Succeeded(
                            result.output ?: DocApprovalOutput.EMPTY,
                            result.events,
                        ),
                    )
                }
            }
        }
    }

    fun reset() {
        submitJob?.cancel()
        submitJob = null
        _uiState.update { it.copy(phase = ExecutionPhase.Idle, showTrace = false) }
    }
}
